contracts/extensions/ERC4907A.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+// ERC721A Contracts v4.1.0
+// Creator: Chiru Labs
+
+pragma solidity ^0.8.4;
+
+import './IERC4907A.sol';
+import '../ERC721A.sol';
+
+/**
+ * @dev ERC4907 compliant extension of ERC721A.
+ *
+ * The ERC4907 standard https://eips.ethereum.org/EIPS/eip-4907[ERC4907] allows
+ * owners and authorized addresses to add a time-limited role
+ * with restricted permissions to ERC721 tokens.
+ */
+abstract contract ERC4907A is ERC721A, IERC4907A {
+    // The bit position of `expires` in packed user info.
+    uint256 private constant _BITPOS_EXPIRES = 160;
+
+    // Mapping from token ID to user info.
+    //
+    // Bits Layout:
+    // - [0..159]   `user`
+    // - [160..223] `expires`
+    mapping(uint256 => uint256) private _packedUserInfo;
+
+    /**
+     * @dev Sets the `user` and `expires` for `tokenId`.
+     * The zero address indicates there is no user.
+     *
+     * Requirements:
+     *
+     * - The caller must own `tokenId` or be an approved operator.
+     */
+    function setUser(
+        uint256 tokenId,
+        address user,
+        uint64 expires
+    ) public virtual {
+        if (!_isApprovedOrOwner(msg.sender, tokenId)) revert SetUserCallerNotOwnerNorApproved();
+
+        _packedUserInfo[tokenId] = (uint256(expires) << _BITPOS_EXPIRES) | uint256(uint160(user));
+
+        emit UpdateUser(tokenId, user, expires);
+    }
+
+    /**
+     * @dev Returns the user address for `tokenId`.
+     * The zero address indicates that there is no user or if the user is expired.
+     */
+    function userOf(uint256 tokenId) public view virtual returns (address) {
+        uint256 packed = _packedUserInfo[tokenId];
+        assembly {
+            // Branchless `packed *= block.timestamp <= expires ? 1 : 0`.
+            packed := mul(
+                packed,
+                // `block.timestamp <= expires ? 1 : 0`.
+                lt(shl(_BITPOS_EXPIRES, timestamp()), packed)
```

Comment:
Is this correct? Let's say timestamp == userInfo.expires. `packed` will end up being greater than `timestamp << 160` because of the bits that represent the user in the first 160 bits of `packed`. Wouldn't you need to zero out the user bits before making a comparison?

Also your comment says <= but here you're using `lt` and not `lte`

Reply:
There is no `lte` in yul. The equivalent is `iszero(gt(..)`.

There is a hidden bitwise trick here:

- If `timestamp()` and `expires` are equal, the existence of an address in the lower 160 bits will allow the `lt` to evaluate to true too.

- In the absence of an address in the lower 160 bits, then the function will return the zero address regardless of the boolean condition.

I've fuzz tested this in Foundry.

Reply:
What if `timestamp()` is slightly greater than `expires`, but the existence of an address in the lower 160 bits causes the less than evaluation to be true?

Reply:
```
<-   upper 96 bits  -><-           lower 160 bits            ->
[     timestamp      ][0x00000000000000...00000000000000000000]  `timestamp << 160`
[      expires       ][                user                   ]  `packed`
```

There are several conditions:
- `timestamp > expires`:
    - If `user != 0`, then `(timestamp << 160) > packed`, which causes `packed` to be multiplied by 0.
    - If `user == 0`, then `(timestamp << 160) > packed`, which causes `packed` to be multiplied by 0.
- `timestamp < expires`:
    - If `user != 0`, then `(timestamp << 160) < packed`, which causes `packed` to be multiplied by 1.
    - If `user == 0`, then `address(uint160(packed))` will give 0.
- `timestamp == expires`:
    - If `user != 0`, then `(timestamp << 160) < packed`, which causes `packed` to be multiplied by 1.
    - If `user == 0`, then `address(uint160(packed))` will give 0.
    
 Note that `expires` and `user` won't overlap, since we have done the proper castings in `setUser`, which implicit masks them.

Reply:
Ok yeah this makes sense, ignore my last question. The logic here makes sense although I'm a bit worried about ppl reading this and not quite understanding how this works.